### PR TITLE
Delete dims

### PR DESCRIPTION
**Pull request by sunalbert**
This PR solves part of #445 . Currently，the reduce functions in Dynet only support reducing along a single dimension. For some operations and cases, we need to reduce along muti dimensions, so I create this PR hoping to provide foundation for muti-dimension reduce.

### CODE REVIEW

File: dynet/dim.h [hunk 1]
```diff
@@ -212,6 +212,35 @@ struct Dim {
     }
   }
   /**
+   * \brief Remove multi-dimensions
+   * \param dims dimensions to be removed, assuming that there is no duplicate elements in it
+   * \param reduce_batch reduce the batch dimension or not
+   */
+  inline void delete_dims(std::vector<unsigned int> dims, bool reduce_batch){
+    DYNET_ARG_CHECK(dims.size() <= nd, "Out of bounds exception in Dim::delete_dims");
```
Comment (neubig) [inline dynet/dim.h]: `vector<bool> deleted_dims(nd, false);`

File: dynet/dim.h [hunk 2]
```diff
@@ -212,6 +212,35 @@ struct Dim {
...
+    for(unsigned int i = 0; i < dims.size(); i++) {
+      DYNET_ARG_CHECK(dims[i] < nd, "Out of bounds exception in Dim::delete_dims");
```
Comment (neubig) [inline dynet/dim.h]: `DYNET_ARG_CHECK(!deleted_dims[dims[i]], "Duplicate dimension in Dim::delete_dims");`
`deleted_dims[dims[i]] = true`

File: dynet/dim.h [hunk 3]
```diff
@@ -212,6 +212,35 @@ struct Dim {
...
+    }
+
+    if(dims.size() == nd){
+        nd = 1;
+        d[0] = 1;
+    }
```
Comment (neubig) [inline dynet/dim.h]: Canonical dynet coding style does `} else {` on one line. Also, `) {` not `){`

File: dynet/dim.h [hunk 4]
```diff
@@ -212,6 +212,35 @@ struct Dim {
...
+    else{
+      int flag = 0;
+      for(unsigned int i = 0; i < nd; i++){
+        if(std::find(dims.begin(), dims.end(), i) == dims.end()){
```
Comment (neubig) [inline dynet/dim.h]: `if(!deleted_dims[i])`
(no need for brackets on a one-line if)

File: dynet/dim.h [hunk 5]
```diff
@@ -212,6 +212,35 @@ struct Dim {
...
+          d[flag++] = d[i];
+        }
+      }
+      nd = flag;
+    }
+
+    if(reduce_batch){
```
Comment (neubig) [inline dynet/dim.h]: Similarly, remove brackets.

**Comment by sunalbert** [comment]
@neubig Thanks for your review and suggestion. I have modified the code and it's style. But the code is little different from what you suggested: just making `deleted_dims[dims[i]] =  true` and no check for duplicate element.

**Comment by neubig** [comment]
Thanks!

**Pull request merged**